src/collapse/vsprof.rs

```diff
@@ -0,0 +1,222 @@
+        let split = if let Some(remainder) = remainder.strip_prefix('"') {
+            remainder.split_once('"')
+        } else {
+            return invalid_data_error!("Unable to parse function name from line:\n{}", line);
+        };
+
+        if let Some((function_name, remainder)) = split {
+            let (number_of_calls, _) = get_next_number(remainder)?;
+
+            let prev_depth = self.stack.len();
+            // There are 3 separate cases to handle regarding the depth:
+            // 1. prev_depth + 1 == depth -> a new function is called, we only need to
+            //    store the function name and the number of times it is called from the
+            //    outer function
+            // 2. prev_depth == depth -> the previous function call was a leaf node, so we
+            //    need to save the current stack and replace the top node with our node
+            //    call
+            // 3. prev_depth > depth -> the previous function call was a leaf node, so we
+            //    need to save the current stack and than we need to pop the top nodes
+            //    until the top node is our parent (i.e. the function which called us)
+            match prev_depth.cmp(&depth) {
+                // Case 1
+                Ordering::Less => {
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+                // Case 2
+                Ordering::Equal => {
+                    self.write_stack(occurences);
+                    self.stack.pop();
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+                // Case 3
+                Ordering::Greater => {
+                    // The Visual Studio profiler outputs the number of times a function is called.
+                    //
+                    // Let's say we have a function `A()` which is called 500 times, and which
+                    // calls a function `B()` 300 times. If we didn't do anything special here,
+                    // this would result in `A()` being assigned 800 samples, giving the impression
+                    // that `A()` only calls `B()` less than 50% of the time, while in fact it is
+                    // called more than 50% of the time.
+                    //
+                    // To prevent this from happening, we instead subtract the number of calls from
+                    // the previous node (in this case `B()`) from the current node (in this case
+                    // `A()`. This leaves `A()` with the correct number of 500 samples.
+                    //
+                    // If the previous number of calls is equal to the current number of calls, we
+                    // don't want to write the current top node, because that would duplicate the
+                    // number of samples for the current node.
```

Comment:
I feel like I'm missing something here. If A is called 500 times, and B is called 500 times _from A_, then won't this result in B not getting included in the flamegraph at all?

Reply:
I tested this, and it seems to work. If I call A 500 times, and B 500 times *from* A (so B actually is called 500 * 500 = 250.000 times), B still shows up on the output. B also has 250.000 samples, so that also seems to be correct.

Reply:
Ah, no, I meant the case where every call to A is also calls B (once each)

Reply:
I checked, this also works.

The reason this works is that `prev_number_of_calls` is initialized to 0, and the top node will never have 0 calls, because than it would not be present in the output. This makes the if statement on line 137 always return true, so the initial top node is always written.

I have updated the comment, hopefully this clears up your confusion.